### PR TITLE
Refactor comments

### LINKED ISSUE #64: Refactor comments serialization
Based on discussion with @dennwc we should refactor the way how comments are serialized. 
This issue is partly related to: https://github.com/bblfsh/cpp-driver/issues/60
Right now, we try to assign comments to existing nodes (as a nested nodes, in ast).
We agreed, that comments can be independent nodes, represented as a flat structure and serialized on the same, top level as ast root.

### PR DESCRIPTION
**Pull request by kuba--**
Signed-off-by: kuba-- <kuba@sourced.tech>
Closes https://github.com/bblfsh/cpp-driver/issues/64

This PR is a proposal is related to the issue https://github.com/bblfsh/cpp-driver/issues/60
but also it's about how we gonna place comments in the tree (native or uast).
So far we loose some comments, or drop them in UAST.
This proposal extracts all comments and serialize all of them, once on top level of AST (under the root).
Thanks to this we we don't loose any comments. 
In my opinion, attaching comments as a flat list under the root (instead of attaching to certain nodes) is easier to process and easier to extract.
My intention was also to keep comments for all transformations (not only in native).
PTAL.

---
This change is [Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)

### CODE REVIEW

File: native/src/main/java/tech/sourced/babelfish/DriverRequest.java [hunk 1]
```diff
@@ -8,36 +8,26 @@
...
             super(e);
         }
     }
+
     public String action;
     public String language;
     public String languageVersion;
     public String content;
-    public String Encoding;
-    @JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (kuba--) [inline native/src/main/java/tech/sourced/babelfish/DriverRequest.java:24]: this property didn't work here!

Comment (kuba--) [reply native/src/main/java/tech/sourced/babelfish/DriverRequest.java:24]: we can also remove all of props. except content (not used).

**Review by dennwc — CHANGES_REQUESTED** [review]
> Reviewed 47 of 47 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: all files reviewed, 2 unresolved discussions (waiting on @kuba--)

___
*[fixtures/asm.cpp.uast, line 17 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8lNxAMzcYf_QfTD8:-Lpn8lNxAMzcYf_QfTD9:bnqhg47) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/fixtures/asm.cpp.uast#L17)):*
> ```
>    LeadingComments: [
>       { '@type': "Comment",
>          '@token': " standard inline assembly",
> ```

Unrelated to your PR, but this caught my attention. Do you now how to we trim those `//`? Why they exists in Native, but not in Annotated?
___
*[native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java, line 229 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8FFd7oHa6yW3ShNb:-Lpn8FFd7oHa6yW3ShNc:b-ze4xh2) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java#L229)):*
> ```Java
>         ArrayList<IASTComment> leadingComments = new ArrayList<>();
>         commentMap.getLeadingMap().values().forEach(leadingComments::addAll);
>         serializeCommentList(leadingComments, "Leading");
> ```

Aren't all the comments "freestanding" now? Maybe we can just add a single `Comments` field on the `TranslationUnit`? Should be a lot easier to work with.

**Review by kuba-- — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: all files reviewed, 2 unresolved discussions (waiting on @dennwc)

___
*[fixtures/asm.cpp.uast, line 17 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8lNxAMzcYf_QfTD8:-LpoqwIR73fEqtEofAkI:bhritsy) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/fixtures/asm.cpp.uast#L17)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Unrelated to your PR, but this caught my attention. Do you now how to we trim those `//`? Why they exists in Native, but not in Annotated?
</blockquote></details>

In native it exists, because `CDT` returns a raw comment text, during transformations we drop it.
In `annotation.go` we have:
```go
	AnnotateType("Comment", MapObj(Obj{
		"Comment": UncommentCLike("text"),
	}, Obj{
		uast.KeyToken: Var("text"),
	}), role.Noop, role.Comment),
```
where `UncommentCLike` comes from _sdk_:
```go
// UncommentCLike removes // and /* */ symbols from a given string variable.
func UncommentCLike(vr string) Op {
	return StringConv(Var(vr), uncomment, comment)
}
```

In normalizer we have:
```go
	MapSemantic("Comment", uast.Comment{}, MapObj(
		Obj{
			"Comment":        CommentText([2]string{"/*", "*/"}, "comm"),
			"IsBlockComment": Bool(true),
		},
		CommentNode(true, "comm", nil),
	)),

	MapSemantic("Comment", uast.Comment{}, MapObj(
		Obj{
			"Comment":        CommentText([2]string{"//", ""}, "comm"),
			"IsBlockComment": Bool(false),
		},
		CommentNode(false, "comm", nil),
	)),
```
___
*[native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java, line 229 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8FFd7oHa6yW3ShNb:-LpornPn-Pbu2O6g7vNW:b-zihnlg) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java#L229)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Aren't all the comments "freestanding" now? Maybe we can just add a single `Comments` field on the `TranslationUnit`? Should be a lot easier to work with.
</blockquote></details>

Yeah, it makes sense. I kept nomenclature from CDT, but after we put all of them to the flat list it can be just comments.
And yes, we can add it to any level. I also tried under `TranslationUnit`, but after I put it under `TranslationUnit` it disappeared from annotated and semantic files.

**Review by kuba-- — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: all files reviewed, 2 unresolved discussions (waiting on @dennwc)

___
*[native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java, line 229 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8FFd7oHa6yW3ShNb:-Lq3-Rn2-ZuSlOIB9fzT:b-9zjeo9) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java#L229)):*
<details><summary><i>Previously, kuba-- (Kuba Podgórski) wrote…</i></summary><blockquote>

Yeah, it makes sense. I kept nomenclature from CDT, but after we put all of them to the flat list it can be just comments.
And yes, we can add it to any level. I also tried under `TranslationUnit`, but after I put it under `TranslationUnit` it disappeared from annotated and semantic files.
</blockquote></details>

@dennwc - Shall we move it outside `ast`, under `TransaltionUnit`?  But how we gonna _catch_ comments in _annotated_ and _semantic_ modes. We only process _ast_ (as far as I noticed).

**Review by dennwc — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: all files reviewed, 1 unresolved discussion (waiting on @kuba--)

___
*[fixtures/asm.cpp.uast, line 17 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8lNxAMzcYf_QfTD8:-Lq33cj63qZnQntHJAz_:b-tuunv5) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/fixtures/asm.cpp.uast#L17)):*
<details><summary><i>Previously, kuba-- (Kuba Podgórski) wrote…</i></summary><blockquote>

In native it exists, because `CDT` returns a raw comment text, during transformations we drop it.
In `annotation.go` we have:
```go
	AnnotateType("Comment", MapObj(Obj{
		"Comment": UncommentCLike("text"),
	}, Obj{
		uast.KeyToken: Var("text"),
	}), role.Noop, role.Comment),
```
where `UncommentCLike` comes from _sdk_:
```go
// UncommentCLike removes // and /* */ symbols from a given string variable.
func UncommentCLike(vr string) Op {
	return StringConv(Var(vr), uncomment, comment)
}
```

In normalizer we have:
```go
	MapSemantic("Comment", uast.Comment{}, MapObj(
		Obj{
			"Comment":        CommentText([2]string{"/*", "*/"}, "comm"),
			"IsBlockComment": Bool(true),
		},
		CommentNode(true, "comm", nil),
	)),

	MapSemantic("Comment", uast.Comment{}, MapObj(
		Obj{
			"Comment":        CommentText([2]string{"//", ""}, "comm"),
			"IsBlockComment": Bool(false),
		},
		CommentNode(false, "comm", nil),
	)),
```
</blockquote></details>

OK, thanks for the clarification. I think it may be inconsistent (in Annotated mode) with other drivers.
___
*[native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java, line 229 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8FFd7oHa6yW3ShNb:-Lq33p3A-O8tnzv0VbhA:b-37b6k9) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java#L229)):*
<details><summary><i>Previously, kuba-- (Kuba Podgórski) wrote…</i></summary><blockquote>

@dennwc - Shall we move it outside `ast`, under `TransaltionUnit`?  But how we gonna _catch_ comments in _annotated_ and _semantic_ modes. We only process _ast_ (as far as I noticed).
</blockquote></details>

Sorry, but aren't they already on the `TranslationUnit`? I was only proposing to merge those `Leading-`, `Trailing-`, `Freestanding-` fields to a single `Comments` lists, and that's it.

**Review by kuba-- — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: 1 of 50 files reviewed, 1 unresolved discussion (waiting on @dennwc)

___
*[native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java, line 229 at r1](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lpn8FFd7oHa6yW3ShNb:-Lq5peYy1qoiwUbY2Ja1:b-896fix) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9fb4446c5d5ca6d297e225f60ae9de149567c5ea/native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java#L229)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Sorry, but aren't they already on the `TranslationUnit`? I was only proposing to merge those `Leading-`, `Trailing-`, `Freestanding-` fields to a single `Comments` lists, and that's it.
</blockquote></details>

Done.

**Review by dennwc — CHANGES_REQUESTED** [review]
> Reviewed 49 of 49 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: all files reviewed, 2 unresolved discussions (waiting on @kuba--)

*[a discussion](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lq5wiTV3cZ4bo_Z_oLy:-Lq5wiTV3cZ4bo_Z_oLz:b-y9b98f) (no related file):*
CI fail. Maybe not all the fixtures committed?

---
___
*[driver/normalizer/normalizer.go, line 144 at r2](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lq5vy1V3a_oW9JX4mxp:-Lq5vy1V3a_oW9JX4mxq:b-cvmz4l) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9eeecd3d91b40a9eeec4f88dbc096ab4b5df7908/driver/normalizer/normalizer.go#L144)):*
> ```Go
> 			// FIXME(juanjux): save this once we've a way
> 			{Name: "Resolved", Op: Any()},
> 			{Name: "Comments", Drop: true, Op: Any()},
> ```

All the comments are now on `CPPASTTranslationUnit`, right? If so, we can safely remove those rules from the normalization code.

**Review by kuba-- — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: 37 of 117 files reviewed, 2 unresolved discussions (waiting on @dennwc and @kuba--)

___
*[driver/normalizer/normalizer.go, line 144 at r2](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lq5vy1V3a_oW9JX4mxp:-Lq6bYy7-l6Aaejwhl2I:bvirjgb) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9eeecd3d91b40a9eeec4f88dbc096ab4b5df7908/driver/normalizer/normalizer.go#L144)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

All the comments are now on `CPPASTTranslationUnit`, right? If so, we can safely remove those rules from the normalization code.
</blockquote></details>

Yeah, I guess.
But I afraid, it may generate some incorrect `.uast` files (with empty objects, but I don't know how to get rid of).

**Review by kuba-- — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: 37 of 117 files reviewed, 2 unresolved discussions (waiting on @dennwc and @kuba--)

___
*[driver/normalizer/normalizer.go, line 144 at r2](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-Lq5vy1V3a_oW9JX4mxp:-Lq6bxXV5R8T-TZOTg7z:b-896fix) ([raw file](https://github.com/bblfsh/cpp-driver/blob/9eeecd3d91b40a9eeec4f88dbc096ab4b5df7908/driver/normalizer/normalizer.go#L144)):*
<details><summary><i>Previously, kuba-- (Kuba Podgórski) wrote…</i></summary><blockquote>

Yeah, I guess.
But I afraid, it may generate some incorrect `.uast` files (with empty objects, but I don't know how to get rid of).
</blockquote></details>

Done.

**Comment by ncordon** [comment]
My concern with this is that this is a behavior we are only introducing in the `cpp-driver`, isn't? What about the other drivers? Even if the other drivers do not loose comments as the C++ one does, it would be desirable that all of them behave as similar as possible.

**Review by ncordon — APPROVED** [review]
> Reviewed 1 of 47 files at r1, 36 of 49 files at r2, 80 of 80 files at r3.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66)* status: all files reviewed, 2 unresolved discussions (waiting on @dennwc and @kuba--)

**Comment by kuba--** [comment]
@ncordon - It's mainly change on _native_ level. So, we didn't modify anything, we just added a new nodes to uast which were missed.
The main difference is that similar comments for instance for go may be assigned to the nodes, but for C/C++ comments will be on top level.

**Comment by ncordon** [comment]
Is there any place in the documentation where we make clear this kind of specific behaviors? If not, any strong opinion where these should be documented?

**Comment by creachadair** [comment]
> Is there any place in the documentation where we make clear this kind of specific behaviors? If not, any strong opinion where these should be documented?

Ideally I think this kind of information should be in the [UAST schema](https://doc.bblf.sh/uast/uast-specification-v2.html). Most of the schema is focused on the encoding details, so I would guess that the [semantic UAST](https://doc.bblf.sh/uast/semantic-uast.html) section is the best location given what we have now. @dennwc what do you think?

Longer term we may want to consider adding a higher-level schema section, to capture meta-properties of the graph above the node level. I don't _think_ we have that yet.

**Review by dennwc — COMMENTED** [review]
@ncordon This change is temporary. The following PRs should move comments closer to other nodes (but not inside), as in most other drivers.

@creachadair Agree, this is not clarified in the schema specifically, although it's implied by a separate `uast:Comment` type.

And as Michael mentioned, in the real graph it may look a bit differently. For example, there is a way to attach comments without changing the schema or affecting the tree traversal for clients. I believe @kuba-- pointed out at some point that the client may want to traverse only "real" AST nodes, not comments.

> Reviewed 2 of 80 files at r3.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-:-Lq8Azt46YF0Qx-Z8Tkv:bnidxxt)* status: all files reviewed, 1 unresolved discussion (waiting on @kuba--)

**Comment by ncordon** [comment]
Okay, thanks for the clarifications, :lgtm:

**Review by dennwc — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/bblfsh/cpp-driver/66#-:-LqKnV6U6AK-Iu1FHbte:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**